### PR TITLE
feat(vitest): add `onTestsRerun` method to global setup context

### LINKED ISSUE #6417: Adding custom watch commands
### Clear and concise description of the problem

We use [testcontainers](https://testcontainers.com/) to run our integration tests with a live PostgreSQL database, in 5 parallel workers

All these DBs must be migrated using Prisma before starting `vitest --watch`, we do this in our global setup:

```typescript
import { PostgreSqlContainer } from "@testcontainers/postgresql";
import { exec as execCb } from "child_process";
import { promisify } from "util";
import type { GlobalSetupContext } from "vitest/node";

const exec = promisify(execCb);

export default async function setup({ provide, config }: GlobalSetupContext) {
  const stopsFunctions: (() => Promise<void>)[] = [];
  const postgresUrls: Record<number, string> = {};

  await prismaGenerate();

  await Promise.all(
    [...Array(config.maxWorkers).keys()].map(async (index) => {
      const postgres = await buildPostgresContainer();
      postgresUrls[index + 1] = postgres.url;
      stopsFunctions.push(postgres.stop);
    }),
  );

  provide("postgresUrls", postgresUrls);

  return async () => {
    await Promise.all(
      stopsFunctions.map(async (stop) => {
        await stop();
      }),
    );
  };
}

const prismaGenerate = async () => {
  console.log("[Prisma] Generating...");
  await exec(`npx prisma generate`);
  console.log("[Prisma] Generated successfully!");
};

const buildPostgresContainer = async () => {
  console.log("[Postgres] Starting...");
  const postgres = await new PostgreSqlContainer("postgres:16")
    .withTmpFs({ "/dev/shm/postgres": "rw,noexec,nosuid,size=65536k" })
    // https://www.postgresql.org/docs/current/non-durability.html
    .withCommand([
      "postgres",
      "-c",
      "fsync=off",
      "-c",
      "synchronous_commit=off",
      "-c",
      "full_page_writes=off",
      "-c",
      "autovacuum=off",
    ])
    .start();

  const url = postgres.getConnectionUri();

  console.log("[Postgres] Migrating...");
  await exec(`DB_URL=${url} npx prisma migrate deploy`);
  console.log("[Postgres] Migrated successfully!");

  return {
    url,
    stop: async () => {
      await postgres.stop();
    },
  };
};

// Type safe access to `provide/inject` methods:
declare module "vitest" {
  export interface ProvidedContext {
    postgresUrls: Record<number, string>;
  }
}
```

If I write a new migration, it is not applied to my live test databases, I must end the Vitest process and start again (in this case, this takes ~30s because of all the containers)

### Suggested solution

Would be nice if we could add a new quick command to the watch menu, such as `m` for triggering migrations inside the databases

### Alternative

We could setup a parallel watching process which runs the migrations every time the migrations folder changes, but that seems flaky

Perhaps related to https://github.com/vitest-dev/vitest/issues/842

### Additional context

_No response_

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by sheremet-va**
### Description

Fixes #6417

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

File: test/watch/test/global-setup-rerun.test.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    watch: true,
+    include: ['math.test.ts'],
+  })
+
+  await vitest.waitForStdout('Waiting for file changes')
+
+  const calls = (globalThis as any).__CALLS as string[]
+  expect(calls).toEqual(['start'])
+
+  editFile(testFile, testFileContent => `${testFileContent}\n\n`)
```
Comment (AriPerkkio) [inline test/watch/test/global-setup-rerun.test.ts:19]: I think we could use `'r'` keypress here instead. Editing files is heavy on CI.

```ts
  vitest.write('r')

  await vitest.waitForStdout('RERUN')
```

Comment (sheremet-va) [reply test/watch/test/global-setup-rerun.test.ts:19]: But we are testing that _watcher_ triggered a callback

Comment (sheremet-va) [reply test/watch/test/global-setup-rerun.test.ts:19]: There should be two tests then, not one

Comment (AriPerkkio) [reply test/watch/test/global-setup-rerun.test.ts:19]: If we are testing file editing here, then yes. If it's just about testing whether rerun calls this callback, then it's not important.

Comment (sheremet-va) [reply test/watch/test/global-setup-rerun.test.ts:19]: We are testing everything that affects this, I don't think about implementation details when testing it. When user does something, something else should happen

**Review by AriPerkkio — DISMISSED** [review]
Looks good to me, I like this idea! 💯 

I wonder if at some point we'll have "plugins" that are registered in the global setup for more hooks. We'll see where this API lands eventually 🤔

**Review by hi-ogawa — APPROVED** [review]
I tested this to make arktype/attest's global setup rerun-able https://github.com/hi-ogawa/reproductions/tree/main/vitest-attest-snapshot It looks working nicely :clap:

**Review by AriPerkkio — APPROVED** [review]

**Review by hi-ogawa — APPROVED** [review]

**Pull request merged**